src/python_testing/TC_TIMESYNC_2_1.py

```diff
@@ -34,40 +34,51 @@ def pics_TC_TIMESYNC_2_1(self) -> list[str]:
 
     @async_test_body
     async def test_TC_TIMESYNC_2_1(self):
+        endpoint = 0
 
-        endpoint = self.user_params.get("endpoint", 0)
+        features = await self.read_single_attribute(dev_ctrl=self.default_controller, node_id=self.dut_node_id,
+                                                    endpoint=endpoint, attribute=Clusters.TimeSynchronization.Attributes.FeatureMap)
+
+        self.supports_time_zone = bool(features & Clusters.TimeSynchronization.Bitmaps.Feature.kTimeZone)
+        self.supports_ntpc = bool(features & Clusters.TimeSynchronization.Bitmaps.Feature.kNTPClient)
+        self.supports_ntps = bool(features & Clusters.TimeSynchronization.Bitmaps.Feature.kNTPServer)
+        self.supports_trusted_time_source = bool(features & Clusters.TimeSynchronization.Bitmaps.Feature.kTimeSyncClient)
+        timesource_attr_id = Clusters.Objects.TimeSynchronization.Attributes.TimeSource.attribute_id
+
+        # Reading attributes from DUT
+        timesync_attr_list = Clusters.Objects.TimeSynchronization.Attributes.AttributeList
+        read_request = await self.default_controller.ReadAttribute(self.dut_node_id, [timesync_attr_list])
+        attributes_list = read_request[0]
+
+        # Extracting gathered attributes dictionary values into a list containing only supported attribute ids
```

Comment:
I'm not totally sure what this is doing - the attribute list should already be a list that only has the supported attribute ids. It's not guaranteed to be ordered.

You should just be able to read the attribute using the above function and check whether the time source is present by doing
if Clusters.TimeSynchronization.Attributes.TimeSource.attribute_id in attribute_list:

Reply:
Thank you,
Using the above method had provided back an embedded dictionary, so had to parse the dictionary to get the attributes list out of it.
However, I have updated this section of the code in TC_TIMESYNC_2_1 test module to now use the get_single_attribute_check_success function from matter_testing_support module which provided the attributes list back in a list, so code is much cleaner now.